modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/StemmerTokenFilterFactoryTests.java

```diff
@@ -97,6 +97,79 @@ public void testPorter2FilterFactory() throws IOException {
+            IndexAnalyzers indexAnalyzers = analysis.indexAnalyzers;
+            NamedAnalyzer analyzer = indexAnalyzers.get("my_plurals");
+            assertThat(create, instanceOf(EnglishPluralStemFilter.class));
+
+            // Check old EnglishMinimalStemmer ("S" stemmer) logic
+            assertAnalyzesTo(analyzer, "phones", new String[]{"phone"});
+            assertAnalyzesTo(analyzer, "horses", new String[]{"horse"});
+            assertAnalyzesTo(analyzer, "cameras", new String[]{"camera"});
+            
+            // The orginal s stemmer gives up on stemming oes words because English has no fixed rule for the stem
+            // (see https://howtospell.co.uk/making-O-words-plural )
+            // This stemmer removes the es but retains e for a small number of exceptions 
+            assertAnalyzesTo(analyzer, "mosquitoes", new String[]{"mosquito"});
+            assertAnalyzesTo(analyzer, "heroes", new String[]{"hero"});
+            // oes exceptions that retain the e.
+            assertAnalyzesTo(analyzer, "shoes", new String[]{"shoe"});
+            assertAnalyzesTo(analyzer, "horseshoes", new String[]{"horseshoe"});
+            assertAnalyzesTo(analyzer, "canoes", new String[]{"canoe"});
+            assertAnalyzesTo(analyzer, "oboes", new String[]{"oboe"});
+
+            // Check improved EnglishPluralStemFilter logic
+            //sses
+            assertAnalyzesTo(analyzer, "dresses", new String[]{"dress"});
+            assertAnalyzesTo(analyzer, "possess", new String[]{"possess"});
+            assertAnalyzesTo(analyzer, "possesses", new String[]{"possess"});
+            // xes
+            assertAnalyzesTo(analyzer, "boxes", new String[]{"box"});
+            assertAnalyzesTo(analyzer, "axes", new String[]{"axe"});
+            //shes
+            assertAnalyzesTo(analyzer, "dishes", new String[]{"dish"});
+            assertAnalyzesTo(analyzer, "washes", new String[]{"wash"});
+            //ees
+            assertAnalyzesTo(analyzer, "employees", new String[]{"employee"});
+            assertAnalyzesTo(analyzer, "bees", new String[]{"bee"});
+            //tch
+            assertAnalyzesTo(analyzer, "watches", new String[]{"watch"});
+            assertAnalyzesTo(analyzer, "itches", new String[]{"itch"});
+            // ies->y but only for length >4
+            assertAnalyzesTo(analyzer, "spies", new String[]{"spy"});
+            assertAnalyzesTo(analyzer, "ties", new String[]{"tie"});
+            assertAnalyzesTo(analyzer, "lies", new String[]{"lie"});
+            assertAnalyzesTo(analyzer, "pies", new String[]{"pie"});
+            assertAnalyzesTo(analyzer, "dies", new String[]{"die"});
+
+            
+            // *CHES - would be good to find a simple rule that solves lunches, churches but doesn't break aches
+            // documenting current behaviour here as a known issue:
+            assertAnalyzesTo(analyzer, "lunches", new String[]{"lunche"});
```

Comment:
Looking at http://www.thefreedictionary.com/words-that-end-in-ches I think that 'avalanche' and 'headache' are the odd ones out here - consider brooch/brooches, branch/branches, couch/couches.  Maybe have a specific rule for `-ache`, and some exceptions?